backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2647,16 +2707,33 @@ public ArtifactEntity getArtifactEntity(Session session, String experimentRunId,
     return new AbstractMap.SimpleEntry<>(artifactEntity.getPath(), uploadId);
   }
 
+  private String buildArtifactPartLockKey(Long artifactId, Long partNumber) {
+    return "artifactId::" + artifactId + "::partNumber::" + partNumber;
+  }
+
   @Override
   public Response commitArtifactPart(CommitArtifactPart request) throws ModelDBException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       ArtifactEntity artifactEntity = getArtifactEntity(session, request.getId(), request.getKey());
-      VersioningUtils.saveOrUpdateArtifactPartEntity(
-          request.getArtifactPart(),
-          session,
-          String.valueOf(artifactEntity.getId()),
-          ArtifactPartEntity.EXP_RUN_ARTIFACT);
-      return Response.newBuilder().build();
+      try (AutoCloseable ignored =
+          acquireWriteLock(
+              buildArtifactPartLockKey(
+                  artifactEntity.getId(), request.getArtifactPart().getPartNumber()))) {
```

Comment:
I thought the reading operation should not an issue for race condition but yes I think it will lead the client to fetch the same part then it will lead to the problem so let me add it for the safe side.